MoreLinq/Memoize.cs

```diff
@@ -0,0 +1,144 @@
+        /// </remarks>
+
+        public static IEnumerable<T> Memoize<T>(this IEnumerable<T> source)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+
+            if (source is ICollection<T>)
+            {
+                return source;
+            }
+
+            return (source as MemoizedEnumerable<T>) ?? new MemoizedEnumerable<T>(source);
+        }
+    }
+
+    internal class MemoizedEnumerable<T> : IEnumerable<T>, IDisposable
+    {
+        private IList<T> cache;
+        private readonly object locker;
+        private readonly IEnumerable<T> source;
+        private IEnumerator<T> sourceEnumerator;
+
+        public MemoizedEnumerable(IEnumerable<T> sequence)
+        {
+            if (sequence == null) throw new ArgumentNullException(nameof(sequence));
+
+            source = sequence;
+            locker = new object();
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            if (cache == null)
+            {
+                lock (locker)
+                {
+                    if (cache == null)
+                    {
+                        var cache = new List<T>(); // for exception safety, allocate then...
+                        sourceEnumerator = source.GetEnumerator(); // (because this can fail)
+                        this.cache = cache; // ...commit to state
+                    }
+                }
+            }
+
+            var index = 0;
+
+            while (true)
+            {
+                bool hasValue;
+                lock (locker)
+                {
+                    if (index < cache.Count)
+                    {
+                        hasValue = true;
+                    }
+                    else if (sourceEnumerator == null)
+                    {
+                        hasValue = false;
+                    }
+                    else if (sourceEnumerator.MoveNext()) // TODO exception safety?
+                    {
+                        hasValue = true;
+                        cache.Add(sourceEnumerator.Current);
+                    }
+                    else
+                    {
+                        hasValue = false;
+                        sourceEnumerator.Dispose();
+                        sourceEnumerator = null;
+                    }
+                }
+
+                if (hasValue)
+                    yield return cache[index];
+                else
+                    break;
+
+                index++;
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public void Dispose()
+        {
+            if (cache == null)
+                return;
+            lock (locker)
+            {
+                if (cache == null)
+                    return;
+                cache = null;
+                sourceEnumerator?.Dispose(); // TODO Test nullability
```

Comment:
This was a to-do comment for @leandromoh as there was no test covering the case where a `MemoizedEnumerable` is disposed when it's in a primed state (i.e. `cache != null`) and the source has been closed (`sourceEnumerator == null`). If the following two tests had been written consistently:

```c#
        [Test]
        public void MemoizeWithDisponseOnEarlyExitTrue()
        {
            using (var xs = new[] { 1, 2 }.AsTestingSequence())
            {
                var memoized = xs.Memoize();
                using ((IDisposable) memoized)
                    memoized.Take(1).Consume();
            }
        }

        [Test]
        public void MemoizeDisponsesAfterSourceIsIteratedEntirely()
        {
            using (var xs = new[] { 1, 2 }.AsTestingSequence())
                xs.Memoize().Consume();
        }
```

then the bug would have been caught. It simply requires writing `MemoizeDisponsesAfterSourceIsIteratedEntirely` as follows:

```c#
        [Test]
        public void MemoizeDisponsesAfterSourceIsIteratedEntirely()
        {
            using (var xs = new[] { 1, 2 }.AsTestingSequence())
            {
                var memoized = xs.Memoize();
                using ((IDisposable) memoized)
                    memoized.Consume();
            }
        }
```